### ISSUE TITLE
Matrix Triangular Solve does not require CUSolver for GPU

### ISSUE THREAD

**Issue by alexggmatthews** [GPflow/GPflow#213]
I was just thinking about this on a walk. See section 3.4.6 of the linked document

https://developer.nvidia.com/sites/default/files/akamai/cuda/files/Misc/mygpu.pdf

Therefore our current bottleneck for SVGP on float32 can be moved to GPU without the difficult changes to the stream executor discussed here:

https://github.com/tensorflow/tensorflow/issues/2217

This is important because removing the MatrixTriangularSolve bottleneck would speed up GPU performance by about a factor of 3.

I guess CUSOLVER is only required for the  Cholesky decomposition, which would be one of our next bottlenecks. 

A look at Tom Nixon's fork confirms the GPU matrix triangular solve implementation only calls CUBLAS not CUSOLVER.

https://github.com/c0g/tensorflow/blob/master/tensorflow/core/kernels/matrix_triangular_solve_op.cc

I wish I had time to work on this right now but I don't unfortunately. Thought everyone should know straight away, anway.

**Comment by alexggmatthews** [comment]
Looking into the Cholesky issue it looks like another route is to use the MAGMA library discussed in the attached doc. Unfortunately this isn't currently supported in TF either.

**Comment by alexggmatthews** [comment]
Given that MatrixTriangularSolve is doable for GPU using only CUBLAS probably so is CholeskyGrad. This would leave just Cholesky itself. It seems like a work around for that might be to implement Cholesky for GPU by hand using CUBLAS as a custom op for GPflow. It can't be more complicated than Cholesky grad. 

Exciting.

**Comment by markvdw** [comment]
Wow this is really exciting. So, just to get things straight, the issue essentially is that the Cholesky decomposition is not in cuBLAS, but it is in cuSOLVER, and that the latter isn't accessible through TensorFlow?

I'm wondering about the feasibility of a manual implementation of Cholesky. My prior is that implementing a numerical routine to be competitive with these highly optimised packages would be very hard.

**Comment by alexggmatthews** [comment]
@markvdw I definitely agree with your last point about Cholesky. I doubt our implementation of Cholesky would be as fast as the library one but it would likely be faster than having to run it on a CPU. The nice thing about this thread is that we can break the thing up into manageable chunks and get gains after each one.

**Comment by alexggmatthews** [comment]
It may not be possible to do MatrixTriangularSolve as a user op because stream executor is not part of the tf includes at run time. I've raised a related issue on the main tf repo here:

https://github.com/tensorflow/tensorflow/issues/4518

**Comment by c0g** [comment]
In mainline

https://github.com/tensorflow/tensorflow/commit/0dace14fd76a6cd975c084fa1f800480ef0aed61

**Comment by dougalsutherland** [comment]
@c0g Thanks for that!

I have a (non-GP) problem where I'd really like to have GPU implementations of Cholesky and CholeskyGrad. I noticed that you [said in May](https://github.com/tensorflow/tensorflow/issues/367#issuecomment-216052421) you had these implemented in your fork using cusolver, but I couldn't find that code. Is it still there somewhere?

**Comment by c0g** [comment]
Yes, it's at [1]. Moved it there to make pull requesting to main TF easier. Warning though, this is tf0.8. I am working on moving these mods to master atm but may take some time.

[1] https://github.com/c0g/tomserflow/tree/master/tensorflow

**Comment by dougalsutherland** [comment]
@c0g Thanks. I can't get tf 0.8 to build on my machine (some bazel problem), and just copying the relevant files into user_ops of master didn't work either as I'm sure you expected. Anyway, if you have it ready on master in the next day or two, I'll definitely use it in an ICLR submission this week...if not, I'd still for sure use it at some point in the future. :)

**Comment by icouckuy** [comment]
@c0g did you had some time to update the fork to tf0.11? We are doing some benchmarks right now on cpu vs gpu, however we find there is little to no difference on our machine.

**Comment by c0g** [comment]
I’m afraid not, and am unlikely to in the future due to time constraints/job.

Tom

**Comment by alexggmatthews** [comment]
@icouckuy A few thoughts: 

GPU is generally a lot faster with float32. 

tf12 will include Tom's ( @c0g ) successful pull request: https://github.com/tensorflow/tensorflow/pull/5010 on MatrixTriangularSolve. 

At the moment Tom's work on GPU kernel's for CholeskyGrad and Cholesky has not made it into TensorFlow main and is only on his fork. If the application you using has one of these tf Ops as its bottleneck then without these you will be CPU bound.

hth,
Alex

**Comment by icouckuy** [comment]
Thanks for the summary. It is a bit surprising that a cholesky kernel does not exist yet for the gpu, as most GP models rely on it. Though GP models using (a small number of) inducing points are probably less affected.

We might take a look at Tom's code in the future as deep GPs don't really make sense without it (the full speed of a GPU will make it easier to test large-scale data sets). A student is working on it, but he has to read up on LVMs and deep GPs first.

**Comment by c0g** [comment]
The work required to get a cholesky kernel, vs a kernel from BLAS is quite a step function so finding someone motivated enough to do it AND get a large change through Google’s process is tough. You can actually run GPFlow (as of a few weeks a go at least) on top of my TF branch with minor changes, if you want to try it out to see if the speed ups are worth it[1].

One idea Alex or James had would be to implemented the cholesky by translating the fortran implementation of LAPACK using level-3 BLAS ops, all of which are wrapped. You won’t get full cuSOLVER performance, but it should definitely be much easier both in terms of the code breadth you have to understand and avoidance of bureaucracy.

[1] Spoiler, they are:

**Comment by icouckuy** [comment]
@c0g should master tomsorflow compile and which bazel version did you use? I'm trying with the latest bazel version and I'm going through several fixes due to the newer bazel version. I arrived now at a point I'm not 100% sure it is bazel or your code being broken atm. Bazel is complaining about undefined dependencies such as:
ERROR: /home/ibcn/icouckuyt/projecten/tomserflow/tensorflow/core/kernels/BUILD:853:1: undeclared inclusion(s) in rule '//tensorflow/core/kernels:cholesky_op':
this rule is missing dependency declarations for the following files included by 'tensorflow/core/kernels/cholesky_op.cc':
  '/home/ibcn/icouckuyt/projecten/tomserflow/tensorflow/core/platform/stream_executor.h'
  '/home/ibcn/icouckuyt/projecten/tomserflow/tensorflow/core/platform/default/stream_executor.h'.

Probably the newer bazel is much more strict in this regard, however, there are no rules for platform that I can simply add to the deps in the BUILD file so fixing your fork is becoming more invasive as you include files from all over the place.

I'm not sure I 100% understand what you mean with a cholesky kernel vs a BLAS kernel. With cholesky kernel you mean cholesky as implemented by nvidia's cusolver and a BLAS kernel would use the wrapper TF BLAS functions which can work on either on CPU or GPU and so the final kernel would be able to work on the CPU/GPU too? I'm just getting into tensorflow so I have no idea why wrapping cusolver would be so much work (but I belief you).

**Comment by alexggmatthews** [comment]
@icouckuy the discussion #2217 describes the difficulty incorporating cusolver in TensorFlow.  I've also discussed this in person with a developer from Google Brain in Mountain View, who agrees that changes to the stream executor are going to be challenging for someone outside that team. hth,
Alex

**Comment by icouckuy** [comment]
Thanks, I indeed read that issue (and all related issues). After thinking it through the problem is pretty clear to me know. As a first step we will be switching to tf master to get the new matrix_triangular_solve and see from there.

I'm not bothered about bureaucrazy - porting the current hack to tf master would be fine for me for now, as well as translating the LAPACK one. For the latter we would need to implement the gradients too.

**Comment by icouckuy** [comment]
We will be implementing the blocked cholesky algorithm from LAPACK, however we will be trying to do it in python using standard TF ops (it seems to be possible I think). As TF probably will use the more efficient CUSolver in the future it will be a temporary solution anyways, no point in spending much effort in it.

If it works out and after some benchmarking, would it be interesting to add to the official GPflow repo as a temporary solution? Until TF has CUSolver support.

**Comment by icouckuy** [comment]
Fyi, a trace file of SVGP with 3000 inducing points on the California housing data set (~20k points) with TF master. Matrix_triangular_solve is no longer to be seen (in TF0.11 it takes a lot of time as you said). While cholesky(_grad) doesn't take much time, copying over data from gpu to cpu seems to take a lot of time if I can belief the trace file. So having a cholesky just to avoid these memcpy will be useful.
![trace_radar_gpu_svgpr_3000](https://cloud.githubusercontent.com/assets/177595/20664695/a3f08596-b55c-11e6-83f5-01db877a7e86.png)

**Comment by dougalsutherland** [comment]
@icouckuy Are you actively working on this now? I'm now at a point where I would really like (for my non-GP application) having a GPU implementation of Cholesky and its gradient to avoid copying on and off the GPU, but will be doing it on small matrices so I don't especially care if it's ultra-efficient.

If you've already started working on this, then I'll just wait a day or two, but if not then I'll give it a shot myself.

**Comment by icouckuy** [comment]
@dougalsutherland we are actively working on it and it is the priority of our master student. However, 2 days is quite short. I'm fairly confident it will take about 2 weeks max, probably sooner but I can not make any promises of course.

I do think it is an excellent starting point for him so if you can wait that long that would be good, if you decide to work because of some deadlines please let me know. Alternatively, you could start on implementing cholesky_grad, we need to integrate this in the TF codebase anyway for GPflow.

**Comment by c0g** [comment]
In Murray’s article on the cholesky gradient the cholgrad_symbolic equation was very competitive with the blocked version for matrices of a few hundred elements, at least on my gtx 970 GPU. That might be an easy win depending on your requirements.

@icouckuy - out of interest, how does your master’s student intend to implement the cholesky in pure TF?

**Comment by alexggmatthews** [comment]
@c0g looking at the paper:

https://arxiv.org/pdf/1602.07527v1.pdf

Blocked chol_rev (which is what tf uses) is about 1.7 x faster than Symbolic chol_rev for N=500 and
Blocked chol_rev (which is what tf uses) is about 3.2 x faster than Symbolic chol_rev for N=4000 . 

He remarks it is somewhat machine dependent.  I guess all his stuff is on CPU.

What is the exact comparison you are doing here, please?

**Comment by c0g** [comment]
I don’t have the code/data with me now, but if my memory serves the overhead of blocked vs. symbolic was higher on the GPU than reported in Murray’s paper. I was not hugely surprised so didn’t investigate further. My memory is that there was little difference or some gain from using symbolic up to a few hundred matrix elements.

Tom

**Comment by alexggmatthews** [comment]
Thanks @c0g  -that's useful to know. Just to confirm, does blocked take over again on GPU above that point?

**Comment by c0g** [comment]
Yep. 

It’s hard coded to 128 in my code which is not optimal, but low on my list of things to work on.

**Comment by dougalsutherland** [comment]
@icouckuy Here's my implementation of unblocked cholesky with TF ops. I couldn't figure out how to get it to work without explicitly knowing the input size at graph-building time, though it's probably possible

```python
def cholesky(X):
    # need to have an explicit shape for X
    shape = X.get_shape()
    shape.assert_has_rank(2)
    m = shape[0].value
    assert m is not None
    assert shape[1].value == m

    A = tf.matrix_band_part(X, -1, 0)
    cond = lambda j, A: j < m
    def body(j, A):
        s = A[j, :j]
        s = A[j, :j]
        s2 = tf.matmul(tf.expand_dims(s, 0), tf.expand_dims(s, 1))[0, 0]
        Ajj = tf.sqrt(A[j, j] - s2)
        
        col = A[j+1:m, j]
        col -= tf.matmul(A[j+1:m, :j], tf.expand_dims(A[j, :j], 1))[:, 0]
        col /= Ajj
        
        A = tf.concat(0, [
            A[:j],
            tf.concat(1, [
                A[j:, :j],
                tf.expand_dims(tf.concat(0, [tf.expand_dims(Ajj, 0), col]), 1),
                A[j:, j+1:]
            ])])
        A.set_shape([m, m])
        return j + 1, A
    return tf.while_loop(
        cond, body, (tf.constant(0, tf.int32), A))[1]
```

This does run on the GPU and gets the right results for the inputs I tried. It's probably much slower than it needs to be, but for my limited current purposes I think this is enough. Hopefully your student can do a better version. :)

Similarly crappy implementation of the gradient coming probably tomorrow.

**Comment by markvdw** [comment]
@dougalsutherland, I think the cholesky gradient is already implemented natively in TensorFlow. Is there any way you could just use that?

**Comment by dougalsutherland** [comment]
@markvdw CholeskyGrad in tensorflow is implemented in eigen only, as far as I can tell, and so won't run on the GPU.

**Comment by icouckuy** [comment]
As far as I know cholesky_grad was added by Alex to TF, it is not part of Eigen. Tom made a gpu version of it in tomserflow and added gpu versions of cholesky, tril and diag.

For our use (GPflow) cholesky_grad needs to be implemented in TF itself, albeit we could do it in python as part of tensorflow/python/ops/linalg_grad.py. But we will see, thanks for this first version we will see if we can improve it and how it performs. Ideally we would port the C++ ops of tomserflow to tf master of course but I figured a python version is easier to start with.

As a sidenote, I was also surprised QR decomposition is not in TF either. We needed it for another project, I would look at proper cusolver support in TF myself but I'm a bit hesitant as Tom has much experience and says it is not easily done without the TF team.

**Comment by dougalsutherland** [comment]
What I meant is that it's implemented using Eigen, and so (I think) CPU-only, in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/cholesky_grad.cc.

**Comment by icouckuy** [comment]
Yep, sorry. misread your sentence.

**Comment by dougalsutherland** [comment]
@icouckuy By the way, I just chanced across this, but there is now a QR op in tensorflow ([docs](https://www.tensorflow.org/versions/master/api_docs/python/math_ops.html#qr)); it was added in https://github.com/tensorflow/tensorflow/commit/715f951eb9ca20fdcef20bb544b74dbe576734da last week.

**Comment by alexggmatthews** [comment]
@dougalsutherland impressed by the Cholesky code - that's neat. can't you get the gradient automatically? Tensorflow has a lot better coverage for numpy like operations these days. I think this would have been quite difficult at the beginning. 

Re my Cholesky grad code in tf main: Yes to reiterate it is currently only CPU. Confusingly they have scripts that do automatic rebasing and have deleted much of the Git history of the file. Tom ( @c0g ) has a GPU version on his unofficial branch.

**Comment by dougalsutherland** [comment]
The automatic gradient doesn't agree with the one that tensorflow computes; I think differentiating through while loops is iffy. The symbolic gradient, though, is easy to write now that tf has triangular matrix solvers on GPU. I'm going to wrap that up into an op soon.

**Comment by dougalsutherland** [comment]
Actually, the gradient for the code I posted before does work: it just gives the lower triangular half, instead of the symmetrized version that `tf.cholesky` gives, which should be equivalent.

I tried to register an explicit symbolic gradient like this:

```python
@tf.RegisterGradient("MyCholesky")
def my_cholesky_grad_sym(L, Lbar):
    def _Phi(A):
        A = tf.matrix_band_part(A, -1, 0)
        return A - tf.matrix_band_part(A, 0, 0) / 2.
    
    P = _Phi(tf.matmul(tf.transpose(L), Lbar))
    return _Phi(tf.matrix_triangular_solve(
        L,
        tf.transpose(tf.matrix_triangular_solve(
            L, P + tf.transpose(P), adjoint=True, lower=True)),
        adjoint=True, lower=True))
```

but it didn't actually end up calling it. Not sure what's going on there, but the version I posted earlier works well enough for my purposes for now. Hopefully @icouckuy's student can make a faster/more flexible version for GPflow's use.

**Comment by vdutor** [comment]
Hello, I'm @icouckuy's student. 

I implemented a blocked version of the Cholesky algorithm. It is based on the LAPACK implementation which uses level-3 BLAS operations (http://www.cs.utexas.edu/~plapack/icpp98/node2.html).

In the plots below, the 'tf' is the Tensorflow Cholesky implementation (i.e. `tf.cholesky`), 'unblocked' is the implementation of @dougalsutherland (posted above) and 'blocked' is my version. In the first plot, the 'blocked' and 'unblocked' implementations run on the GPU and the 'tf' runs on the CPU. In the second plot all the implementations are running on the CPU.
![cholesky_tf_master_gpu](https://cloud.githubusercontent.com/assets/13780846/20962360/f34979e2-bca3-11e6-98fb-882a16a536e6.png)
![cholesky_tf_master_cpu](https://cloud.githubusercontent.com/assets/13780846/20962359/f347bd46-bca3-11e6-9f59-ec13fb89719b.png)
According to the second plot, the blocked version is faster than Tenserflow's `tf.cholesky`, even if they run both on CPU. I find this quite surprising. I already double checked my code and the code of Google's Eigen implementation (https://android.googlesource.com/platform/external/eigen/+/jb-mr1.1-dev/Eigen/src/Cholesky/LLT.h) for a possible explanation. Without success, so if anyone finds a mistake or explanation for this, please let me know.

The benchmarks were run on a machine with an Intel Xeon CPU and a Nvidia 1080 GPU.

The `cholesky_blocked` code can be found here: https://github.com/vdutor/cholesky. Any suggestions and improvements are welcome.

Is there any interest in temporarily integrating this in GPflow until Tensorflow comes up with GPU support for the Cholesky?

**Comment by jameshensman** [comment]
My machine doesn't replicate this! (intel i7, 4 cores (8 hyperthread), no GPU), using tf version 0.11.0rc0.

![updated_cholesky](https://cloud.githubusercontent.com/assets/154383/20965029/f13febc8-bc6b-11e6-899f-5e5c2c4ce256.png)

**Comment by jameshensman** [comment]
and here are the results on my gpu (intel xeon, K40). Looks like there is some GPU-based advantage.

![updated_cholesky](https://cloud.githubusercontent.com/assets/154383/20965616/715d4308-bc6e-11e6-9a96-d59ca8ef44f0.png)

**Comment by vdutor** [comment]
I ran the benchmarks on a server configured with Tensorflow master. On my own laptop (i7 CPU, Nvidia GTX 960M GPU), I get similar results as the plots you just uploaded.

**Comment by icouckuy** [comment]
I'm thinking some parallelization advantages. We ran it on a server with 2 CPUs (20 physical cores)

**Comment by jameshensman** [comment]
...and here are the results for my GPU machine with the precision set to float32...
![updated_cholesky](https://cloud.githubusercontent.com/assets/154383/20965720/f840534c-bc6e-11e6-82da-2397100ee3f3.png)

**Comment by jameshensman** [comment]
I think there's a good case for including this in GPflow. I wonder if backpropagation is also faster in the blocked algorithm? could you benchmark on a real datset, including hyper-parameter learning? this is easy to do: just inherit from the gpr.GPR class and overwrite build_likelihood to include your code, something like

```python
import GPflow
from cholesky import cholesky_blocked

class GPR_blocked(GPflow.gpr.GPR):
    def build_likelihood(self):
        K = self.kern.K(self.X) + eye(tf.shape(self.X)[0]) * self.likelihood.variance
        L = cholesky_blocked(K)
        m = self.mean_function(self.X)
        return multivariate_normal(self.Y, m, L)

m = GPR_blocked(X, Y, ...)
```

If we were to include it in GPflow, we should probably make it optional using the config file.

**Comment by icouckuy** [comment]
That's our next  step, rerun our GPflow benchmarks with the new cholesky, though not sure if GPR will be in it (SVGP and LVM will be). Our data set is the california housing data, about 30k points I belief.

For modifying GPflow we were thinking to replace all tf.cholesky calls (there are not that many) with a util.cholesky() returning the right op (either based on a config option or the current device context).

Notice that the pure TF cholesky generates a lot of TF ops. While not a problem for just training a model, it yields unwieldy trace files (94mb for just one cholesky of a 3k matrix). We will be putting it into a name_scope but I think this only works for tensorboard, not trace files.

**Comment by alexggmatthews** [comment]
Hi everyone, 
This is cool. I have quite a few questions, though. 

I' am concerned about the memory footprint of this as @icouckuy mentions. It seems it might result in potentially quite a large footprint from doing this at the computational graph level, just as there is for instance in backprop through time. 

In the CPU case have you considered a hybrid algorithm that calls the Cholesky Op for the blocks?

**Comment by icouckuy** [comment]
The memory footprint is certainly something to consider. However, at the moment we are mostly concerned with automatic differentiation which does not seem to work or is just very slow - seems to get stuck in an infinite loop. Hopefully it is a bug we can solve. Otherwise, an option is to use the symbolic gradient in pure TF but I don't see how to 'glue' this to the cholesky op so TF knows what gradient to call when attaching a TF optimizer to the graph. Hints are welcome, if it is at all possible. The TF documentation is very sparse - non-existing - on python user ops (tensorflow/python/user_opts/).

It would be great if we could define the cholesky op as a kind of python user op using linalg methods so the footprint in the computational graph (and the trace files) stays small, just like gradients are in tensorflow/python/ops/linalg_grad.py. Anyway, this seem an awfully lot of work and we might just as well port it to a C++ op as suggested at the start of this issue :-)

We haven't considered calling tf.cholesky for the blocks. That's certainly an option though our goal was (quickly) creating a GPU op and avoiding some memcpy's in the process.

**Comment by alexggmatthews** [comment]
@icouckuy thanks for the update. I'm currently at nips #290 but since there seems to be a lot on interest in this thread I will bump it up my to do list. I hope to have another chat to Tom @c0g about the C++ options when I get back.

**Comment by icouckuy** [comment]
@alexggmatthews good to hear. We decided to port our cholesky algorithm to a C++ kernel. There are just too many problems with memory footprint and coupling gradients to it. I think it should still be a pretty forward port using the wrapped BLAS interface of stream.

**Comment by dougalsutherland** [comment]
@icouckuy @vdutor Any progress on a proper C++ implementation?

**Comment by icouckuy** [comment]
@dougalsutherland sorry, I didn't see your question until now. Unfortunately, we don't have a C++ implementation yet. During the work our objectives changed a bit.

We do have improved the python TF version a bit, made it recursive instead of the tf.while, and the ajoint gradients work with it. We will post some results later.

**Comment by dougalsutherland** [comment]
@icouckuy Cool. I keep running into the need for this in slightly different settings, but I see that @vdutor updated the repository so I'll just try that next time it comes up.

**Comment by vdutor** [comment]
I made some benchmarks of the current implementation, that might be interesting for @dougalsutherland or others. The previous implementation made use of a `tf.while` construction, which led to a kind of deadlock during the gradient calculation. We revised the implementation into a recursive one. The gradient calculation of this recursive implementation now works fine. The summary of the benchmarks: [chol_benchmarks.pdf](https://github.com/GPflow/GPflow/files/932746/chol_benchmarks.pdf).

**Comment by jameshensman** [comment]
@vdutor great stuff, thanks for sharing. 

I think it could be improved very slightly by having the same y-axis for the GPU and CPU plots :)

**Comment by alexggmatthews** [comment]
So TensorFlow now has GPU implementations of Cholesky and CholeskyGrad on GPU. It is on master waiting to be released but those who build their on tf can benefit from it now.

**Comment by c0g** [comment]
*awesome*

**Comment by icouckuy** [comment]
The work on cuSolver and cholesky GPU op started a  while back in april:
https://github.com/tensorflow/tensorflow/commit/72c023d3967a3218cd3d830ce6e57f7c4d87a18c

It is already in TF from v1.2.0.

**Comment by alexggmatthews** [comment]
My apologies for the lack of clarity. Cholesky gpu is in already. Cholesky Grad GPU is awaiting release but is on master. The Linked PR covers Cholesky only- to get the full speed up in GPflow you need both.

**Comment by icouckuy** [comment]
Yes, sorry. I was only talking about cholesky :-) It's great news.

It seems there are still some other ops (Diag, DiagPart, etc.) running on cpu. Don't know how they will influence the performance, while not expensive they might induce cpu<->gpu transfers.

**Comment by alexggmatthews** [comment]
Yes good spot. I suspect they will slow things due to memory operations.

I hope to put them on GPU too. I think that MatrixDiag and MatrixDiagPart have GPU implementations. They works slightly differently but I suspect that isn't a big problem.

**Comment by alexggmatthews** [comment]
This last point is sorted in #445

**Comment by nknudde** [comment]
Great news!
